### PR TITLE
Feature: Api item to string

### PR DESCRIPTION
**Pull request by MartynasZilinskas**

### CODE REVIEW

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts [hunk 1]
```diff
@@ -80,9 +128,10 @@ export abstract class BasePlugin<TKind = Contracts.ApiItemDto> implements Plugin
             .Text(parsedReturnType)
             .GetOutput();
 
+        //FIXME: Reference
         // pluginResult.UsedReferences = parsedReturnType.References;
         return pluginResult;
     }
 
-    public abstract Render(data: PluginOptions<TKind>): PluginResult;
+    public abstract Render(data: PluginOptions, apiItem: TKind): PluginResult;
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/abstractions/base-plugin.ts]: Rename to `options`.

File: packages/ts-docs-gen/src/abstractions/container-plugin.ts [hunk 2]
```diff
@@ -14,82 +15,87 @@ export interface ContainerMembersKindsGroup {
...
-    protected RenderMembersGroups(options: PluginOptions<TKind>, list: ContainerMembersKindsGroup[]): PluginResultData {
-        const membersReferences = this.getItemsReferenceByKind(options.ExtractedData, list, options.ApiItem.Members);
+    // TODO: Remake options.
+    protected RenderMembersGroups(
+        options: PluginOptions,
+        list: ContainerMembersKindsGroup[],
+        members: ApiDefinitions[],
+        includeHr = true,
+        headingLevel = 2,
+        other = true
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/abstractions/container-plugin.ts]: Rename `other`.  Maybe `shouldRenderUnsupportedMembers` or `shouldRenderUnlistedMembers`.

File: packages/ts-docs-gen/src/api-items/api-definition-base.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import { ApiItemReference } from "../contracts/api-item-reference";
+import { ApiDefinitions } from "./api-definition-list";
+import { ApiTypes } from "./api-type-list";
+
+/**
+ * Base definition class with helper functions.
+ */
+export abstract class ApiDefinitionBase<TKind extends Contracts.ApiBaseItemDto = Contracts.ApiBaseItemDto> extends BaseApiItemClass<TKind>
+    implements SerializedApiDefinition<TKind> {
+
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/api-definition-base.ts:16]: Remove empty line.

File: packages/ts-docs-gen/src/api-items/api-definition-base.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    protected MembersToText(members: ApiDefinitions[], tab: number = 0): string[] {
+        return Helpers.Flatten(
+            members.map(x => x.ToText().map(y => `${GeneratorHelpers.Tab(tab)}${y}`))
+        );
+    }
+
+    protected SerializedTypeToString(apiType: ApiTypes | undefined): string {
+        if (apiType == null) {
+            return "???";
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/api-definition-base.ts]: Add `TODO`.

File: packages/ts-docs-gen/src/api-items/api-type-base.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+export abstract class ApiTypeBase<TKind extends Contracts.ApiBaseType> extends BaseApiItemClass<TKind> implements SerializedApiType<TKind> {
+    protected GetSerializedApiDefinition(referenceId: string): ApiDefinitions {
+        const apiItem = this.ExtractedData.Registry[referenceId];
+        return GeneratorHelpers.SerializeApiDefinition(this.ExtractedData, apiItem, { Alias: "", Id: referenceId });
+    }
+
+    protected SerializedTypeToString(apiType: SerializedApiType | undefined): string {
+        if (apiType == null) {
+            return "???";
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/api-type-base.ts]: Add `TODO`.

File: packages/ts-docs-gen/src/api-items/definitions/api-construct.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { Contracts } from "ts-extractor";
+import { ApiCallable } from "../api-callable";
+
+export class ApiConstruct extends ApiCallable<Contracts.ApiConstructDto> {
+    public ToText(): string[] {
+        return [
+            `new ${this.CallableToString(undefined)};`
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/definitions/api-construct.ts]: ApiConstruct can have return type.

File: packages/ts-docs-gen/src/api-items/types/api-type-definition.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+export type TypeDefinitions = Contracts.TypeLiteralType |
+    Contracts.MappedType |
+    Contracts.FunctionTypeType |
+    Contracts.ThisType |
+    Contracts.ConstructorType;
+
+export class ApiTypeDefinition<TKind extends Contracts.ApiReferenceBaseType = TypeDefinitions> extends ApiTypeReferenceBase<TKind> {
+    public ToText(): string[] {
+        if (this.ReferenceItem == null) {
+            return ["???"];
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/types/api-type-definition.ts]: Add `TODO`.

File: packages/ts-docs-gen/src/api-items/types/api-type-query.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Example: `typeof Foo`
+ */
+export class ApiTypeQuery extends ApiTypeReferenceBase<Contracts.TypeQueryType> {
+    public ToText(): string[] {
+        let name: string;
+        if (this.ReferenceItem != null) {
+            name = this.ReferenceItem.Data.Name;
+        } else {
+            name = "???";
```
Comment (DeividasBakanas) [inline packages/ts-docs-gen/src/api-items/types/api-type-query.ts]: Add `TODO`.

**Review by DeividasBakanas — CHANGES_REQUESTED** [review]

File: packages/ts-docs-gen/src/api-items/api-definition-base.ts (hunk 3 above)
Comment (MartynasZilinskas) [reply packages/ts-docs-gen/src/api-items/api-definition-base.ts:16]: It's for readability.

File: packages/ts-docs-gen/src/api-items/definitions/api-construct.ts (hunk 6 above)
Comment (MartynasZilinskas) [reply packages/ts-docs-gen/src/api-items/definitions/api-construct.ts]: Nice catch.

**Review by DeividasBakanas — APPROVED** [review]

File: packages/ts-docs-gen/src/abstractions/base-api-item.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import { SerializedApiItem } from "../contracts/serialized-api-item";
+
+export abstract class BaseApiItemClass<TKind> implements SerializedApiItem<TKind> {
+    constructor(private extractedData: ExtractDto, private apiItem: TKind) { }
+
+    protected get ExtractedData(): ExtractDto {
+        return this.extractedData;
+    }
+
+    public get Data(): TKind {
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/base-api-item.ts]: Why rename `apiItem` to `Data`?

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts [hunk 10]
```diff
@@ -27,29 +71,33 @@ export abstract class BasePlugin<TKind = Contracts.ApiItemDto> implements Plugin
             // ConstraintType
             let constraintType: string;
             if (typeParameter.ConstraintType != null) {
-                constraintType = GeneratorHelpers.ApiTypeToString(extractedData, typeParameter.ConstraintType);
+                constraintType = typeParameter.ConstraintType.ToText().join(" ");
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/base-plugin.ts]: -> ToInlineText()

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts [hunk 11]
```diff
@@ -27,29 +71,33 @@ export abstract class BasePlugin<TKind = Contracts.ApiItemDto> implements Plugin
...
                 });
             } else {
+                // FIXME: References.
                 defaultType = "";
                 // defaultType = { References: [], Text: "" };
             }
 
             return [
-                typeParameter.Name,
+                typeParameter.Data.Name,
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/base-plugin.ts]: Add `Name` property to `typeParameter`

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts [hunk 12]
```diff
@@ -65,13 +113,13 @@ export abstract class BasePlugin<TKind = Contracts.ApiItemDto> implements Plugin
         return pluginResult;
     }
 
-    protected RenderType(extractedData: ExtractDto, type: Contracts.ApiType | undefined): PluginResultData | undefined {
+    protected RenderType(type: SerializedApiType | undefined): PluginResultData | undefined {
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/base-plugin.ts]: `SerializeApiType` -> `SerializeApiTypes`

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts [hunk 13]
```diff
@@ -80,9 +128,10 @@ export abstract class BasePlugin<TKind = Contracts.ApiItemDto> implements Plugin
             .Text(parsedReturnType)
             .GetOutput();
 
+        //FIXME: Reference
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/base-plugin.ts]: Space before `FIXME`

File: packages/ts-docs-gen/src/abstractions/container-plugin.ts [hunk 14]
```diff
@@ -14,82 +15,101 @@ export interface ContainerMembersKindsGroup {
...
+                MembersList: membersList
             });
         }
 
         return result;
     }
 
-    protected RenderMembersGroups(options: PluginOptions<TKind>, list: ContainerMembersKindsGroup[]): PluginResultData {
-        const membersReferences = this.getItemsReferenceByKind(options.ExtractedData, list, options.ApiItem.Members);
+    // TODO: Remake options.
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/abstractions/container-plugin.ts]: Remove comment

File: packages/ts-docs-gen/src/api-items/api-definition-base.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public get Reference(): ApiItemReference {
+        return this.reference;
+    }
+
+    private parentItem: ApiDefinitions | undefined;
+
+    public get ParentItem(): ApiDefinitions | undefined {
+        if (this.Data.ParentId != null) {
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/api-definition-base.ts]: Make `ParentItem` lazier. Second call should not call `this.GetSerializedApiDefinition(parentReference);`

File: packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+            }
+        }
+        return this.type;
+    }
+
+    private resolveType(): Contracts.ApiType | undefined {
+        let type: Contracts.ApiType | undefined;
+        if (this.Data.ApiKind === Contracts.ApiItemKinds.GetAccessor) {
+            // GetAccessor
+
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts]: Remove empty line.

File: packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            type = this.Data.Type;
+        } else if (this.Data.Parameter != null) {
+            // SetAccessor
+
+            const apiParameter = this.ExtractedData.Registry[this.Data.Parameter.Ids[0]] as Contracts.ApiParameterDto;
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts]: Remove empty line.

File: packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+                type = apiParameter.Type;
+            }
+        }
+
+        return type;
+    }
+
+    public ToText(): string[] {
+        const name = this.Reference.Alias || this.Data.Name;
+        const abstract = this.Data.IsAbstract ? " abstract" : "";
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts]: I need a dollar, dollar, dollar is what I need.

File: packages/ts-docs-gen/src/api-items/definitions/api-class-method.ts [hunk 19]
```diff
@@ -0,0 +1,23 @@
+import { Contracts } from "ts-extractor";
+import { ApiCallable } from "../api-callable";
+
+export class ApiClassMethod extends ApiCallable<Contracts.ApiClassMethodDto> {
+    public ToText(): string[] {
+        const name = this.Reference.Alias || this.Data.Name;
+
+        const optional = this.Data.IsOptional ? "?" : "";
+        const abstract = this.Data.IsAbstract ? " abstract" : "";
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-class-method.ts]: $

File: packages/ts-docs-gen/src/api-items/definitions/api-class-method.ts [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    public ToText(): string[] {
+        const name = this.Reference.Alias || this.Data.Name;
+
+        const optional = this.Data.IsOptional ? "?" : "";
+        const abstract = this.Data.IsAbstract ? " abstract" : "";
+        const async = this.Data.IsAsync ? " async" : "";
+        const $static = this.Data.IsStatic ? " static" : "";
+        const functionHeader = this.CallableToString(`${optional}: `);
+
+        const access = this.Data.AccessModifier;
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-class-method.ts]: access -> accessModifier

File: packages/ts-docs-gen/src/api-items/definitions/api-class-property.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
+import { Contracts } from "ts-extractor";
+import { ApiDefinitionWithType } from "../api-definition-with-type";
+
+export class ApiClassProperty extends ApiDefinitionWithType<Contracts.ApiClassPropertyDto> {
+    public ToText(): string[] {
+        const name = this.Reference.Alias || this.Data.Name;
+
+        const optional = this.Data.IsOptional ? "?" : "";
+        const readOnly = this.Data.IsReadonly ? " readonly" : "";
+        const abstract = this.Data.IsAbstract ? " abstract" : "";
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-class-property.ts]: $

File: packages/ts-docs-gen/src/api-items/definitions/api-index.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+import { ApiParameter } from "./api-parameter";
+import { ApiDefinitionWithType } from "../api-definition-with-type";
+
+export class ApiIndex extends ApiDefinitionWithType<Contracts.ApiIndexDto> {
+    private parameter: ApiParameter;
+
+    public get Parameter(): ApiParameter {
+        if (this.parameter == null) {
+            const apiItem = this.ExtractedData.Registry[this.Data.Parameter] as Contracts.ApiParameterDto;
+            this.parameter = new ApiParameter(this.ExtractedData, apiItem, { Alias: "", Id: this.Data.Parameter });
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-index.ts]: "" -> apiItem.Name

File: packages/ts-docs-gen/src/api-items/definitions/api-mapped.ts [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+import { ApiDefinitionWithType } from "../api-definition-with-type";
+import { GeneratorHelpers } from "../../generator-helpers";
+
+export class ApiMapped extends ApiDefinitionWithType<Contracts.ApiMappedDto> {
+    private typeParameter: ApiTypeParameter | undefined;
+
+    public get TypeParameter(): ApiTypeParameter | undefined {
+        if (this.typeParameter == null && this.Data.TypeParameter != null) {
+            const apiItem = this.ExtractedData.Registry[this.Data.TypeParameter] as Contracts.ApiTypeParameterDto;
+            this.typeParameter = new ApiTypeParameter(this.ExtractedData, apiItem, { Alias: "", Id: this.Data.TypeParameter });
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/definitions/api-mapped.ts]: "" -> apiItem.Name

File: packages/ts-docs-gen/src/api-items/types/api-type-indexed-access.ts [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+import { SerializedApiType } from "../../contracts/serialized-api-item";
+
+/**
+ * Example: `Foo[T]`
+ */
+export class ApiIndexedAccess extends ApiTypeBase<Contracts.IndexedAccessType> {
+    constructor(extractedData: ExtractDto, apiItem: Contracts.IndexedAccessType) {
+        super(extractedData, apiItem);
+
+        this.objectType = GeneratorHelpers.SerializeApiType(this.ExtractedData, this.Data.ObjectType);
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/types/api-type-indexed-access.ts]: Make it lazy.

File: packages/ts-docs-gen/src/api-items/types/api-type-indexed-access.ts [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+        this.indexType = GeneratorHelpers.SerializeApiType(this.ExtractedData, this.Data.IndexType);
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/types/api-type-indexed-access.ts]: Make it lazy.

File: packages/ts-docs-gen/src/api-items/types/api-type-operator.ts [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+import { ApiTypes } from "../api-type-list";
+
+/**
+ * Example: `keyof Foo`
+ */
+export class ApiTypeOperator extends ApiTypeBase<Contracts.TypeOperatorType> {
+    constructor(extractedData: ExtractDto, apiItem: Contracts.TypeOperatorType) {
+        super(extractedData, apiItem);
+
+        this.type = GeneratorHelpers.SerializeApiType(this.ExtractedData, this.Data.Type);
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/types/api-type-operator.ts]: Make it lazy.

File: packages/ts-docs-gen/src/api-items/types/api-type-union-or-intersection.ts [hunk 27]
```diff
@@ -0,0 +1,15 @@
+import { Contracts } from "ts-extractor";
+import { ApiTypeMembersBase } from "../api-type-members-base";
+
+/**
+ * Example: `Foo | Bar` or `Foo & Bar`.
```
Comment (DovydasNavickas) [inline packages/ts-docs-gen/src/api-items/types/api-type-union-or-intersection.ts]:
```
/**
* Examples:
* `Foo | Bar`
* `Foo & Bar`
*/
```

**Review by DovydasNavickas — CHANGES_REQUESTED** [review]

File: packages/ts-docs-gen/src/abstractions/base-plugin.ts (hunk 12 above)
Comment (MartynasZilinskas) [reply packages/ts-docs-gen/src/abstractions/base-plugin.ts]: It's named `ApiTypes`.

File: packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts (hunk 18 above)
Comment (MartynasZilinskas) [reply packages/ts-docs-gen/src/api-items/definitions/api-accessor.ts]: Hey hey 🎵

**Review by DovydasNavickas — APPROVED** [review]

**Pull request merged**